### PR TITLE
Add connector for elasticsearch

### PR DESCRIPTION
**Pull request by livelxw**
#395

### CODE REVIEW

File: es/src/it/scala/monix/connect/es/ElasticsearchSinkSuite.scala [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+import org.scalatest.matchers.should.Matchers
+
+class ElasticsearchSinkSuite extends AnyFlatSpecLike with Fixture with Matchers with BeforeAndAfterEach {
+  import com.sksamuel.elastic4s.http.ElasticDsl._
+  override def beforeEach(): Unit = {
+    super.beforeEach()
+    elasticClient.execute(deleteIndex("*")).runSyncUnsafe()
+  }
+  "ElasticsearchSink" should "update a element in batches" in {
+    val updateRequests = Gen.listOfN(10, genUpdateRequest).sample.get
```
Comment (paualarco) [inline es/src/it/scala/monix/connect/es/ElasticsearchSinkSuite.scala:17]: could we separate each the test in the three steps (as it is done in other connectors)

//given
(values perepared for the test)

//when 
(actions)

//then
(checks)

File: es/src/it/scala/monix/connect/es/ElasticsearchSinkSuite.scala [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+      .bufferTumbling(5)
+      .consumeWith(ElasticsearchSink.bulk)
+      .runSyncUnsafe()
+    val r = updateRequests.map { request =>
+      getById(request.indexAndType.index, request.id)
+        .map(_.sourceAsString)
+        .runSyncUnsafe()
+    }
+    r should contain theSameElementsAs updateRequests.flatMap(_.documentSource)
+  }
```
Comment (paualarco) [inline es/src/it/scala/monix/connect/es/ElasticsearchSinkSuite.scala:29]: could you add more tests cases like failed ones or consuming empty observables that will test that the `onError` and `onComplete` differently? :)

also would be good to add some tests (examples) for the other type of bulk requests like `Deletes` and `Index`

File: es/src/main/scala/monix/connect/es/Elasticsearch.scala [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+  import com.sksamuel.elastic4s.http.ElasticDsl._
+
+  /**
+    * Execute bulk requests
+    *
+    * @param requests some [[BulkCompatibleRequest]]
+    * @param client an implicit instance of a [[ElasticClient]]
+    * @return a [[Task]] with [[Response]] with [[BulkResponse]]
+    */
+  def bulkRequest(requests: Seq[BulkCompatibleRequest])(
```
Comment (paualarco) [inline es/src/main/scala/monix/connect/es/Elasticsearch.scala:21]: what do you think abut adding more methods that will serve other type of single requests? 
(count, singleUpdate, refresh, search, merge, delete...)

File: es/src/main/scala/monix/connect/es/package.scala [hunk 4]
```diff
@@ -0,0 +1,14 @@
+package monix.connect
+
+import com.sksamuel.elastic4s.http._
+import monix.eval.Task
+
+package object es {
+
+  implicit val taskFunctor: Functor[Task] = new Functor[Task] {
```
Comment (paualarco) [inline es/src/main/scala/monix/connect/es/package.scala:8]: should both variables be private `private[s3]`?

File: project/Dependencies.scala [hunk 5]
```diff
@@ -97,4 +97,9 @@ object Dependencies {
     "com.google.cloud" % "google-cloud-nio" % Versions.GCNio % IntegrationTest,
     "commons-io" % "commons-io" % "2.6" % Test
   ) ++ testDependencies(hasIt = true)
+
+  val ES = Seq(
+    "com.sksamuel.elastic4s" %% "elastic4s-core" % "6.7.8",
```
Comment (paualarco) [inline project/Dependencies.scala:102]: use latest version? `7.9.1`

File: es/src/main/scala/monix/connect/es/Elasticsearch.scala (hunk 3 above)
Comment (paualarco) [reply es/src/main/scala/monix/connect/es/Elasticsearch.scala:21]: maybe we could also have an `ElasticSearchSource` or `ElasticSearchObservable` that will return _responses_ in form of `Observable`, not sure how to do so. 
It it seems like elastic4s provides a reactive api but that is backed with `Akka` and therefore we would need to create an `ActorSystem`, probably too heavy...

**Review by paualarco — COMMENTED** [review]
@livelxw thanks for this PR, looking good! :)
have left few comments / questions, it might need some more time to review since I am not very familiarised with `elasticsearch`

File: es/src/it/scala/monix/connect/es/ElasticsearchSinkSuite.scala (hunk 1 above)
Comment (livelxw) [reply es/src/it/scala/monix/connect/es/ElasticsearchSinkSuite.scala:17]: ok

File: es/src/it/scala/monix/connect/es/ElasticsearchSinkSuite.scala (hunk 2 above)
Comment (livelxw) [reply es/src/it/scala/monix/connect/es/ElasticsearchSinkSuite.scala:29]: ok

File: es/src/main/scala/monix/connect/es/Elasticsearch.scala (hunk 3 above)
Comment (livelxw) [reply es/src/main/scala/monix/connect/es/Elasticsearch.scala:21]: OK. I'll add them.

File: project/Dependencies.scala (hunk 5 above)
Comment (livelxw) [reply project/Dependencies.scala:102]: Version `7.x.x` is incompatibility with version `6.x.x` and older versions. I have no idea to support different ES versions. Any advice?

Comment (paualarco) [reply project/Dependencies.scala:102]: well, a different submodule for each version might be a solution but I prefer to start with a single one and that would rather be the latest. what do you think?

Comment (livelxw) [reply project/Dependencies.scala:102]: OK

File: es/src/main/scala/monix/connect/es/package.scala [hunk 6]
```diff
@@ -1,14 +1,15 @@
...
 
 package object es {
 
-  implicit val taskFunctor: Functor[Task] = new Functor[Task] {
+  @InternalApi private[es] implicit val taskFunctor: Functor[Task] = new Functor[Task] {
     override def map[A, B](fa: Task[A])(f: A => B): Task[B] = fa.map(f)
   }
 
-  implicit val taskExecutor: Executor[Task] = (client: HttpClient, request: ElasticRequest) =>
+  @InternalApi private[es] implicit val taskExecutor: Executor[Task] = (client: HttpClient, request: ElasticRequest) =>
```
Comment (paualarco) [inline es/src/main/scala/monix/connect/es/package.scala:13]: maybe we could just refer to the existing instances from `elastic4s`? https://github.com/sksamuel/elastic4s/blob/master/elastic4s-effect-monix/src/main/scala/com/sksamuel/elastic4s/monix/instances/TaskInstances.scala

File: es/src/main/scala/monix/connect/es/package.scala [hunk 7]
```diff
@@ -0,0 +1,15 @@
+package monix.connect
+
+import com.sksamuel.elastic4s._
+import monix.eval.Task
+import monix.execution.internal.InternalApi
+
+package object es {
+
+  @InternalApi private[es] implicit val taskFunctor: Functor[Task] = new Functor[Task] {
+    override def map[A, B](fa: Task[A])(f: A => B): Task[B] = fa.map(f)
```
Comment (paualarco) [inline es/src/main/scala/monix/connect/es/package.scala:10]: maybe we could just refer to the existing instances from elastic4s? https://github.com/sksamuel/elastic4s/blob/master/elastic4s-effect-monix/src/main/scala/com/sksamuel/elastic4s/monix/instances/TaskInstances.scala

**Review by paualarco — COMMENTED** [review]
Did left some small comments, will try to review the `ElasticsearchSource` implementation whenever I find some more time! 
Would be good to see more examples of how to search using the `Source`, maybe you could add them in the tests and will also be useful for documentation as well :D

File: es/src/it/scala/monix/connect/es/ElasticsearchSuite.scala [hunk 8]
```diff
@@ -0,0 +1,193 @@
...
+    // then
+    intercept[NoSuchElementException] {
+      client
+        .execute(getIndex(indexName))
+        .runSyncUnsafe()
+        .result
+    }
+  }
+
+  it should "execute a single count request" in {
```
[the remark this answers is not in the document]
Comment (livelxw) [reply es/src/it/scala/monix/connect/es/ElasticsearchSuite.scala:180]: OK

File: es/src/it/scala/monix/connect/es/ElasticsearchSourceSuite.scala [hunk 9]
```diff
@@ -0,0 +1,77 @@
...
+    // when
+    Elasticsearch.bulkRequest(updateRequests).runSyncUnsafe()
+    Elasticsearch.refresh(index).runSyncUnsafe()
+
+    // then
+    val r = source.map(_.id).toListL.runSyncUnsafe()
+    r should contain theSameElementsAs updateRequests.map(_.id).distinct
+  }
+
+  it should "fails when es error occurs" in {
```
[the remark this answers is not in the document]
Comment (livelxw) [reply es/src/it/scala/monix/connect/es/ElasticsearchSourceSuite.scala:48]: OK

File: es/src/main/scala/monix/connect/es/ElasticsearchSource.scala [hunk 10]
```diff
@@ -0,0 +1,99 @@
...
+import scala.collection.mutable
+
+/**
+  *  A pre-built [[monix.reactive.Observable]] implementation that
+  *  publishes documents using an elasticsearch scroll cursor.
+  *
+  * @param settings settings for how documents are queried
+  * @param client an implicit instance of a [[ElasticClient]]
+  */
+@InternalApi private[es] class ElasticsearchSource(settings: SearchSettings)(implicit client: ElasticClient)
```
[the remark this answers is not in the document]
Comment (livelxw) [reply es/src/main/scala/monix/connect/es/ElasticsearchSource.scala:21]: OK

File: es/src/main/scala/monix/connect/es/ElasticsearchSource.scala [hunk 11]
```diff
@@ -0,0 +1,99 @@
...
+  extends Observable[SearchHit] {
+
+  import com.sksamuel.elastic4s.ElasticDsl._
+
+  require(settings.search.keepAlive.isDefined, "The SearchRequest must have a scroll defined (a keep alive time)")
+  private var scrollId: String = _
+  // Parse the keep alive setting out of the original query.
+  private val keepAlive = settings.search.keepAlive.getOrElse("1m")
```
[the remark this answers is not in the document]
Comment (livelxw) [reply es/src/main/scala/monix/connect/es/ElasticsearchSource.scala:29]: OK

File: es/src/main/scala/monix/connect/es/package.scala (hunk 6 above)
Comment (livelxw) [reply es/src/main/scala/monix/connect/es/package.scala:13]: The version of `monix` in `elastic4s` dependencies maybe the old version. Will this cause incompatibilities and conflicts?

Comment (paualarco) [reply es/src/main/scala/monix/connect/es/package.scala:13]: ah, good point! let's leave it as is then :)

File: es/src/it/scala/monix/connect/es/ElasticsearchSourceSuite.scala [hunk 12]
```diff
@@ -15,26 +16,33 @@ class ElasticsearchSourceSuite extends AnyFlatSpecLike with Fixture with Matcher
...
 
     // then
-    source.isFailure shouldBe true
+    val r = source
+      .bufferTimedAndCounted(100.millis, 100)
+      .map(_.map(_.id))
+      .toListL
+      .runSyncUnsafe()
+      .flatten
+    r should contain theSameElementsAs updateRequests.map(_.id).distinct
```
Comment (paualarco) [inline es/src/it/scala/monix/connect/es/ElasticsearchSourceSuite.scala:37]: is `.distinct` necessary? the source only reads the same element once, right?

Comment (livelxw) [reply es/src/it/scala/monix/connect/es/ElasticsearchSourceSuite.scala:37]: The `id` of generated requests might be duplicated.

File: build.sbt [hunk 13]
```diff
@@ -151,6 +151,8 @@ lazy val s3 = monixConnector("s3", Dependencies.S3)
 
 lazy val gcs = monixConnector("gcs", Dependencies.GCS)
 
+lazy val es =  monixConnector("es", Dependencies.ES)
```
Comment (paualarco) [inline build.sbt:154]: we'd need to add that as a dependency to main module _monnix-connect_ in both `.depends(...)` and `aggregate(...)`

File: docker-compose.yml [hunk 14]
```diff
@@ -38,4 +38,16 @@ services:
   redis:
     image: redis
     ports:
-      - 6379:6379
\ No newline at end of file
+      - 6379:6379
+
+  elasticsearch:
```
Comment (paualarco) [inline docker-compose.yml:43]: it needs to be added to the `start-dependencies.sh`, if not the pipeline won't pass.
`docker-compose -f docker-compose.yml up -d minio localstack redis mongo elasticsearch`

Comment (paualarco) [reply docker-compose.yml:43]: althought we could also run all the docker-compose services without need of specifying each of them, so in the future we'll avoid new contributions to worry about this .
it would just be `docker-compose -f docker-compose.yml up -d`

File: project/Dependencies.scala [hunk 15]
```diff
@@ -24,6 +24,7 @@ object Dependencies {
     val Scalacheck = "1.14.0"
     val Mockito = "1.15.0"
     val GCNio = "0.121.2"
+    val Elastic4s = "7.9.1"
```
Comment (paualarco) [inline project/Dependencies.scala:27]: I think we could write some nice documentation already! 
Se we would just need to create a new `elasticsearch.md` under `./docs` folder. 
And then we will have to do minimal changes (which you can take other connectors as reference) on `website/sidebars.json`, `website/pages/en/index.js` and `website/i18/en.json`. 

To see if the changes were correctly applied:
```
sbt docs/docusaurusCreateSite
cd website/
yarn start
```

Comment (paualarco) [reply project/Dependencies.scala:27]: Just before writing documentation, I wanted to let you know that some of the existing apis are getting moved to a more purely functional design, using _cats effect_ to acquire and release resources (in this case it would be `ElasticClient`).

Maybe we can go with this implementation as a first version of the elasticsearch connector, and make those changes if we see it convenient in the future. I leave that up to you.

If you are interested in doing it now, you can follow the this PR as an example: https://github.com/monix/monix-connect/pull/379

**Review by paualarco — COMMENTED** [review]
Before writing documentation, I wanted to let you know that some of the existing apis are getting moved to a more purely functional design, using cats effect to acquire and release resources (in this case it would be ElasticClient).

Maybe we can go with this implementation as a first version of the elasticsearch connector, and make those changes if we see that it is convenient in the future. I leave that up to you.

In case you are interested in doing it now, you can follow the this PR as an example: #379

Would be very easy to add a dedicated web documentation page for `elasticsearch`.
Se we would just need to create a new _elasticsearch.md_ under _./docs_ folder.
And then we will have to do minimal changes (which you can take other connectors as reference) on _website/sidebars.json_, _website/pages/en/index.js_ and _website/i18/en.json_.

To build the website locally and see if the changes were correctly applied:
install yarn
```
sbt docs/docusaurusCreateSite
cd website/
yarn start
```

File: project/Dependencies.scala (hunk 15 above)
Comment (livelxw) [reply project/Dependencies.scala:27]: OK. I'll try this next commit.

Comment (paualarco) [reply project/Dependencies.scala:27]: Cool!

File: es/src/main/scala/monix/connect/es/ElasticsearchSource.scala [hunk 16]
```diff
@@ -0,0 +1,95 @@
...
+  *
+  * @param request [[SearchRequest]]
+  * @param client an implicit instance of a [[ElasticClient]]
+  */
+@InternalApi private[es] class ElasticsearchSource(request: SearchRequest)(implicit client: ElasticClient)
+  extends Observable[SearchHit] {
+
+  import com.sksamuel.elastic4s.ElasticDsl._
+
+  private var scrollId: String = _
```
Comment (paualarco) [inline es/src/main/scala/monix/connect/es/ElasticsearchSource.scala:25]: could that be an `Option[String]` monad that initially is set to `None`?

File: es/src/main/scala/monix/connect/es/ElasticsearchSource.scala [hunk 17]
```diff
@@ -0,0 +1,95 @@
...
+  private def maybeFetch(buffer: mutable.Queue[SearchHit], sub: Subscriber[SearchHit]): Task[Unit] = {
+    if (buffer.isEmpty)
+      fetch(buffer, sub)
+    else
+      Task.unit
+  }
+
+  // if no fetch is in progress then fire one
+  private def fetch(buffer: mutable.Queue[SearchHit], sub: Subscriber[SearchHit]): Task[Unit] = {
+    Option(scrollId) match {
```
Comment (paualarco) [inline es/src/main/scala/monix/connect/es/ElasticsearchSource.scala:85]: maybe removing `maybeFetch` and adding it here would make it more readable?

```
 if (buffer.isEmpty) 
      scrollId match {
        case None => client.execute(request.keepAlive(keepAlive)).map(populateHandler(_, buffer, sub))
        case Some(id) => client.execute(searchScroll(id).keepAlive(keepAlive)).map(populateHandler(_, buffer, sub))
      }
    else Task.unit
```

File: es/src/main/scala/monix/connect/es/ElasticsearchSource.scala [hunk 18]
```diff
@@ -0,0 +1,95 @@
...
+
+  private var scrollId: String = _
+  // Parse the keep alive setting out of the original query.
+  private val keepAlive = request.keepAlive.getOrElse("1m")
+
+  override def unsafeSubscribeFn(subscriber: Subscriber[SearchHit]): Cancelable = {
+    runLoop(mutable.Queue.empty, subscriber).runToFuture(subscriber.scheduler)
+  }
+
+  private def runLoop(buffer: mutable.Queue[SearchHit], sub: Subscriber[SearchHit]): Task[Unit] = {
```
Comment (paualarco) [inline es/src/main/scala/monix/connect/es/ElasticsearchSource.scala:33]: don't really like dealing with mutable variables... thought I think this implementation is good and safe

what do you think on merging run loop and fast loop together?

like:
```
  private def fastLoop(buffer: mutable.Queue[SearchHit], sub: Subscriber[SearchHit]): Task[Unit] = {
    for {
      _ <- fetch(buffer, sub)
      ack <- {
        if (buffer.nonEmpty) Task.deferFuture(sub.onNext(buffer.dequeue))
        else Task.now(Ack.Stop)
      }
    } yield {
      ack match {
        case Ack.Continue =>
          fastLoop(buffer, sub)
        case Ack.Stop =>
          sub.onComplete()
          Task.unit
      }
    }
  } 
```

Comment (livelxw) [reply es/src/main/scala/monix/connect/es/ElasticsearchSource.scala:33]: OK

File: es/src/main/scala/monix/connect/es/ElasticsearchSource.scala (hunk 17 above)
Comment (livelxw) [reply es/src/main/scala/monix/connect/es/ElasticsearchSource.scala:85]: OK

File: es/src/main/scala/monix/connect/es/ElasticsearchSource.scala (hunk 16 above)
Comment (livelxw) [reply es/src/main/scala/monix/connect/es/ElasticsearchSource.scala:25]: Sure

File: es/src/main/scala/monix/connect/es/Elasticsearch.scala [hunk 19]
```diff
@@ -1,141 +1,211 @@
...
 /**
-  * Object for managing the Elasticsearch
+  * Singleton object provides builders for [[Elasticsearch]].
+  *
+  * ===Example===
+  *{{{
+  * import com.sksamuel.elastic4s.ElasticDsl._
+  * import com.sksamuel.elastic4s._
+  *
+  * val esResource = Elasticsearch.create("http://localhost:9200")
```
Comment (paualarco) [inline es/src/main/scala/monix/connect/es/Elasticsearch.scala:44]: very nice to see examples like this one! :D

Comment (livelxw) [reply es/src/main/scala/monix/connect/es/Elasticsearch.scala:44]: Coming soon

**Comment by livelxw** [comment]
@paualarco Maybe there is some thing wrong about github-actions
```
The `set-env` command is deprecated and will be disabled on November 16th. Please upgrade to using Environment Files. For more information see: https://github.blog/changelog/2020-10-01-github-actions-deprecating-set-env-and-add-path-commands/
```
```
The `add-path` command is deprecated and will be disabled on November 16th. Please upgrade to using Environment Files. For more information see: https://github.blog/changelog/2020-10-01-github-actions-deprecating-set-env-and-add-path-commands/
```

**Comment by paualarco** [comment]
@livelxw have rebased master and pushed a commit to add the docs to the webpage, will look into the ci pipeline issue now!

File: build.sbt [hunk 20]
```diff
@@ -151,6 +151,8 @@ lazy val s3 = monixConnector("s3", Dependencies.S3).aggregate(awsAuth).dependsOn
...
```
Comment (paualarco) [inline build.sbt:154]: I think it would be better to rename the module name to `elasticsearch`, since it would also end up being the artifact name, and es is not very representative, agree?

Comment (livelxw) [reply build.sbt:154]: Yes

**Comment by paualarco** [comment]
@livelxw the pipeline is fixed :)

**Comment by paualarco** [comment]
Do you plan to add more changes? Otherwise, I am happy to merge already :)

**Comment by livelxw** [comment]
@paualarco It can be merged I think. I'll provide some more PRs if needed.
